presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static java.util.Objects.requireNonNull;
+
+public class MergeSortProcessor
```

Comment:
Have you considered reusing `com.facebook.presto.operator.MergeHashSort`? It seems to be providing similar functionality and a lot of concepts are repeating (e.g: `PageWithPosition`). `MergeHashSort` currently uses hash value for sorting, but it should be easy to extend it so it takes `MergeSortComparator/PageComparator` and possibly `PageSuppliers` instead of `Iterators<Page>`